### ISSUE TITLE
Perceived differences between jengaroot and magic

### ISSUE THREAD

**Issue by bsansouci** [bsansouci/jengaroot#3]
This is more of tracker issue, or an open discussion, about the best way to modify jengaroot to fit our needs.
Ultimately, we want one jbuild file (or package.json) at the top level (project root) that describe everything needed to build the whole project. We might have jbuild files in the dependency folder (if they get publish with it) that would describe how to build that dependency. We also want to have all build artifacts in `topLevelDir/_build` instead of mingled with the source code.

Right now the problem seems to be that jengaroot crawls the directory tree and checks for jbuild files in every folder, which are then used to generate rules which cannot have targets outside of the current scope directory (where the jbuild file is). This is enforced by `check_for_non_local_rules`. I'm guessing the reason for this restriction is to avoid circular dependencies between build artifacts causing the build process to never reach a fixed point. Though this makes sense with the current way jengaroot works, we need to generate all artifacts inside a `_build` folder.

My guess as a way to approach this is to basically modify [executables_rules](https://github.com/bsansouci/jengaroot/blob/9249e8bfa4b85230bc097a0393726d648dfbea33/jenga/root.ml#L4098) to use recursively build from a `src` folder using [native_compile_ml](https://github.com/bsansouci/jengaroot/blob/9249e8bfa4b85230bc097a0393726d648dfbea33/jenga/root.ml#L2838) or [byte_compile_ml](https://github.com/bsansouci/jengaroot/blob/9249e8bfa4b85230bc097a0393726d648dfbea33/jenga/root.ml#L2894) and then modify [link_executable](https://github.com/bsansouci/jengaroot/blob/9249e8bfa4b85230bc097a0393726d648dfbea33/jenga/root.ml#L4045) to link against `topLevelDir/_build/...` build artifacts instead of `topLevelDir/src/...`.

This would break the symmetry of jengaroot, where each sub-folder decides on how to compile its code, and its parent doesn't need to know anything about it (or jenga will error if it can't find modules etc...). I'm not sure why we'd want that "purity", because it prevents us from moving the build artifacts inside a specific folder.

**Comment by Nick-Chapman** [comment]
The rational for [check_for_non_local_rules] is so jenga can locate the rule for a demanded target. If the rule could come be generated in any directory then jenga has to generate the rules for the entire tree before it can build anything.

We considered relaxing the restriction to allow the per-directory rules to contains targets for the current directory & any sub-directory. This would be quite useful in a number of circumstances, and would require jenga only generating rules for each ancestor directory (of a target), up to the root, when locating the rule for target. But we never did this.

**Comment by Nick-Chapman** [comment]
Regarding your goal to generate all rules inside a single _build directory: again we did try this at one time, and got it 95% working, but we never switched to this model. There were some corner issues which were a pain for us (I can't remember the details) and also it has a performance implication in terms of additional rules to copy sources into the _build tree. However, what we had may be useful for you.

Arseniy (aalekseyev) is going to see if it is possible to recover this code from the bit-rot state in which it is currently languishing.

**Comment by bsansouci** [comment]
hey @Nick-Chapman, thanks for explaining!
I'd love to see why that was so hard. I got this tiny jengaroot.ml file https://gist.github.com/bsansouci/bb0c54de315a585e88fe03bbec8e214f working. All it does is compile one file with all of its artifacts put in _build. I used absolute paths and made sure to _generate_ the artifacts in the right place to avoid the cost of copying anything over. 
It's a tiny example of how I'd go about it.

Is there any reason why you decided to go with having a jbuild file in every subfolder? I can see how this recursive property makes certain things easier, but it certainly makes other things really hard.

Also so far I've only been able to run jenga in a folder with a jengaroot.ml. Is there any way to take all of those build rules and compile them into an executable that I can run in any folder that has a jbuild file?

**Comment by Nick-Chapman** [comment]
Hi Ben,

> working. All it does is compile one file with all of its artifacts put in _build. I used absolute paths and made sure to generate the artifacts in the right place to avoid the cost of copying anything over.  It's a tiny example of how I'd go about it.

I took a look at your example. One thing to notice: every target is setup by a rule defined in the scheme for the directory in which that target will be created. As I mentioned above, this is how is must be so jenga can find the rule for any demanded target.

So there is no need to use absolute paths in your rules. Actions are run in the local directory (of the scheme that setup their rule), and so action command strings normally refer to file basenames [Path.basename] (i.e. for the "-o" arg) and files not in the local directory can be located using a relative(dot-dot) path [Path.reach_from ~dir].

Using relative paths for files under the jengaroot has the benefit that the entire tree can be relocated without need to rerun any build-actions.


> Is there any reason why you decided to go with having a jbuild file in every subfolder?

Normally jbuild files are managed by the developers of the directories in which they reside. Unlike the monolithic jengaroot which is managed by a small group of `build-engineers'. (To give us a fancy title!). Our experience of monolithic files which are widely edited is not good. It results in frequent merge conflicts and it plays badly with the notion of code ownership and code review obligations. Also, jenga would still have the notion of per-directory rules, even without jbuild files in each directory.


> Also so far I've only been able to run jenga in a folder with a jengaroot.ml.

You can run jenga in any sub-directory; by default jenga will build the .DEFAULT alias in that subdir. The build-root is determined as the (nearest) ancestor directory containing a jengaroot.ml or jenga.conf.


> Is there any way to take all of those build rules and compile them into an executable that I can run in any folder that has a jbuild file?

There is no support for compiling rules into an executable - but I'm completely sure I understand what you want to do.

**Comment by aalekseyev** [comment]
Actually we do have an example of how to build a jenga binary statically linked with the rules. This is somehow omitted from the external release, but the code for the binary is rather trivial:

    open Core.Std
    open Command.Let_syntax

    module J = Jenga_lib

    let () =
      Command.run
        (Command.basic'
           ~summary:("build specified targets")
           [%map_open
             let config = J.Cmd_build.config_param
             in
             fun () ->
               let root_dir = J.Path.Abs.create (Core.Std.Sys.getcwd ()) in
               J.Run.main' ~root_dir (Env (fun () -> Root.env)) config
           ])

Here, Root module is your jengaroot.ml. One caveat here is that top-level code of Root module can't do some things that depend on jenga having been initialized, such as [Path.to_absolute_string].

That said, we never needed this for anything but jengaroot profiling.

**Comment by aalekseyev** [comment]
I should expand a bit on how we approached the problem on putting the build artifacts into a separate directory. 

As Nick said, we would set up jenga rules (one rule per file) to copy everything over from the source directory to the _build directory. Then we would rely on existing rules working with (almost) no change within the _build directory.

The main reason we did it like that instead of doing what you're doing is to avoid having to change all existing rules. Indeed, some of the rules are library-specific and live in their jbuild files, and we were reluctant to introduce this complexity there.

One problem with this approach is that it makes your build system much more sensitive to imprecise dependency specifications: if a rule fails to mention a dependency on a source file (or worse, if it depends on directory listing only), the copying rule will not get triggered and your build fails. This is in some ways a good thing (we found a bunch of missing dependencies this way), but it's not a reliable check and makes build fail non-deterministically.

I just looked at this patch and it does need some work to make it work with our current version of jengaroot (a lot of merge conflicts). If you want to consider this approach, I might try and extract something that works, but if you are almost writing your own jengaroot from scratch anyway, then your original approach of setting up rules in the correct place from the outset is probably better.

**Comment by bsansouci** [comment]
Thanks a lot for your explanations!
What you said @Nick-Chapman about relative dir makes sense, I understand how to do it now.
I think we kind of want to reuse as much of Jengaroot as possible, because it does a LOT of stuff. But we can't rely on copying artifacts over. It's also slower than generating them directly into the right place.
That said we might be able to get away with a stricter folder structure. Each project contains:
- a `_build` folder that contains all of the artifacts generated when building. Deleting this folder results in a _clean_ build. This folder has the exact same shape as the root folder: it has a node_modules and a src folder, but in place of the source code there are build artifacts. This 1-1 mapping makes it easier to reason about.
- a `src` folder that might be a flat (for simplicity for now) directory with all of the source code. There's a property in the jbuild file that'll indicate which file is the main one from which to start crawling the dependencies.
- a `node_modules` folder which is where the dependencies go. They might have a jbuild file themselves explaining how to build them, but they might not.

The project has _one_ jbuild which describes the versions of all the dependencies, but it won't say which file needs what. We'll need to extensively use `ocamldep` to check what the dependency tree is, if there's any dependency missing we'll use some package manager X to pull them inside node_modules and continue on. We also need to warn when there are missing deps from jbuild and too many deps in `node_modules`.

Basically I'd love to reuse as many rules as possible, because there's a lot of good stuff in there. We just need to adapt it somehow.

**Comment by bsansouci** [comment]
Oh something I just noticed, jenga makes `.jenga` and `.jenga.tmp` folders where the nearest `jengaroot` is. What exactly's put in there? What happens if build jengaroot to be one executable and then run it in multiple projects? Any chances of collisions that can go badly?

**Comment by Nick-Chapman** [comment]
Embedded (by ocaml_plugin) in jenga.exe are .cmi's needed for compiling the
jengaroot.ml - these are unpacked from the executable into .jenga. Also
ocaml_plugin caches the compilation of jengaroot.ml in .jenga making it
quicker on subsequent runs of jenga.exe (so long as jengaroot.ml is
unchanged). Dont understand your second question. Don't know of any
collisions that can go badly.

.jenga.tmp - If I remember, this is not part of the core jenga
functionality, but rather the big jengaroot.ml sets this up as the value of
env-var TMPDIR.


On 30 April 2016 at 22:47, Benjamin San Souci <notifications@github.com>
wrote:

> Oh something I just noticed, jenga makes .jenga and .jenga.tmp folders
> where the nearest jengaroot is. What exactly's put in there? What happens
> if build jengaroot to be one executable and then run it in multiple
> projects? Any chances of collisions that can go badly?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/bsansouci/jengaroot/issues/3#issuecomment-215997303>
>



-- 
Nick Chapman <nic@amadido.co.uk>
Beira Cottage, The Phygtle, Chalfont St. Peter, Bucks, SL9 0JT, UK.
+44 1494 876885 (home)  +44 7779 121533 (mobile)

**Comment by bsansouci** [comment]
Oh if .jenga is just jenga related things (and not related to the project that its building) then it's all good.

**Comment by bsansouci** [comment]
Well I pushed some changes that move all build artifacts into _build where the jengaroot.ml file is. That _build file should really be moved to be at the top level of the project but that's just details. I'll close this issue and will make new ones if necessary.
Thanks for your help!
